Проект 2. Анализ данных на HeadHunter

Импортируем библиотеки и прописываем параметры подключения и функцию для подключения

In [1]:
# Импорт необходимых библиотек
import psycopg2
import pandas as pd

#Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# функция подключения
connection = psycopg2.connect(
   dbname=DBNAME,
   user=USER,
   host=HOST,
   password=PASSWORD,
   port=PORT
)


Задание 3.1. Сколько вакансий есть в базе?

In [2]:
query = f'''select 
           count(id)
           from public.vacancies
                   '''

# Получаем результаты запроса сразу в виде датафрейма
df_vacan = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_vacan)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\3410585366.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacan = pd.read_sql_query(query, connection)


,count
0,49197


Вывод по заданию №3.1. Всего в базе данных содержатся сведения о 49197 вакансиях.

Задание 3.2 Сколько работодателей в базе?

In [3]:
query = f'''select 
           count(name)
           from public.employers
                  '''

# Получаем результаты запроса сразу в виде датафрейма
df_emloyer= pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_emloyer)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\3392389849.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_emloyer= pd.read_sql_query(query, connection)


,count
0,23501


Вывод по заданию №3.2. Всего в базе данных содержатся сведения о 23501 работодателе. Таким образом, в среднем на каждого работодателя приходится 2.09 опубликованных вакансии.

Задание 3.3 Сколько  регионов в базе?

In [4]:
query = f'''select 
           count(id)
           from public.areas
                   '''

# Получаем результаты запроса сразу в виде датафрейма
df_areas = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_areas)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\3696323919.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_areas = pd.read_sql_query(query, connection)


,count
0,1362


Вывод по заданию 3.3. В базе содержится информация о 1362 регионах, что косвенно указывает на то, что в среднем в одном регионе опубликовано по 36 вакансий.

Задание 3.4. Сколько сфер деятельности в базе?

In [5]:
query = f'''select 
           count(id)
           from public.industries
                   '''

# Получаем результаты запроса сразу в виде датафрейма
df_industries = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_industries)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\2495538331.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_industries = pd.read_sql_query(query, connection)


,count
0,294


Вывод по заданию 3.4. В базе содержатся сведения о вакансиях по 294 сферам деятельности, что может косвенно указывать на то, что усредно, в одной сфере деятельности работатет порядке 80 организаций.

Задание 4.1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area). 
Выберите пятёрку лидеров по количеству вакансий

In [6]:
#Задаем ограничение вывода, так как нас интересует только 5 лидеров
n = 5
query = f'''select
           a.name, 
           count(v.id)
           from public.areas a
               join public.vacancies v on a.id=v.area_id
           group by a.name
           order by 2 desc
           limit {n}
        '''

# Получаем результаты запроса сразу в виде датафрейма
df_lider_vacan = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_lider_vacan)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\3122290118.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_lider_vacan = pd.read_sql_query(query, connection)


,name,count
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892


Вывод по заданию 4.1. Москва является лидером по количеству опубликованных вакансий, что обусловлено двумя факторами: в Москве больше проживает людей и в Москве больше работодателей (больше крупных компаний, которым требуются работники разных специальностей). В целом, по результатам вывода можно отметить, что чем больше проживает людей в регионе, тем больше в нем опубликовано вакансий.

Задание 4.2. Посмотрим на зарплаты. У какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой?

In [ ]:
query = f'''select
           count(id)
           from vacancies
           where salary_from is not null or salary_to is not null
                '''

# Получаем результаты запроса сразу в виде датафрейма
df_salary_info = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_salary_info)

C:\Users\user\AppData\Local\Temp\ipykernel_13680\2514001704.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_salary_info = pd.read_sql_query(query, connection)


,count
0,24073


Вывод по заданию 4.2. Заработная плата указана для 24073 вакансий, что говорит о том, что для каждой второй вакансии указан уровень заработной платы для ориентира, а для половины вакансий такая информация отсутсвует.

Задание 4.3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого числа.

In [ ]:
#4.3.1. Найдим среднее значение для нижней границы зарплатной вилки
query = f'''select
           round(avg(salary_from),0)
           from vacancies
        '''

# Получаем результаты запроса сразу в виде датафрейма
df_salary_f_avg = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_salary_f_avg)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\2090336102.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_salary_f_avg = pd.read_sql_query(query, connection)


,round
0,71065.0


In [8]:
#4.3.2. Найдим среднее значение для верхней границы зарплатной вилки
query = f'''select
           round(avg(salary_to),0)
           from vacancies
         '''

# Получаем результаты запроса сразу в виде датафрейма
df_salary_t_avg = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_salary_t_avg)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\2492445979.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_salary_t_avg = pd.read_sql_query(query, connection)


,round
0,110537.0


Вывод по заданию 4.3. Верхняя средняя граница заработной платы составляет 110537 руб., нижняя средняя граница заработной платы составляет 71065 руб. Таким образом, коэффициент размаха между средней нижней границей и средней верхней границей составляет 1,5

Задание 4.4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Какая пара находится на втором месте по популярности?

In [9]:
#Заданим ограничение вывода, которое по условию задания составляет 2.
n = 2
query = f'''select
           schedule,
           employment,
           count(id)
           from vacancies
           group by 1, 2
           order by 3 desc
           limit {n}
        '''

# Получаем результаты запроса сразу в виде датафрейма
df_schedule = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_schedule)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\2670430906.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_schedule = pd.read_sql_query(query, connection)


,schedule,employment,count
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802


Вывод по заданию 4.4. Удаленная работа при полной занятости вторая пара по популярности в вакансиях.

Задание 4.5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [10]:
query = f'''select
           experience,
           count(id)
           from vacancies
           group by 1
           order by 2 asc
        '''

# Получаем результаты запроса сразу в виде датафрейма
df_experience = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_experience)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\2644944142.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_experience = pd.read_sql_query(query, connection)


,experience,count
0,Более 6 лет,1337
1,Нет опыта,7197
2,От 3 до 6 лет,14511
3,От 1 года до 3 лет,26152


Вывод по заданию 4.5. Результаты анализа вакансий указывают на то, что самые востребуемые специалисты с опытом работы от 1 года до 3 лет. При этом, компании активнее ищут специалистов не имеющих опыта в 5 раз чаще, чем специалистов с опытом работы более 6 лет.

Задание 5.1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [11]:
#Указываем ограничитель вывода запроса.
n = 5
query = f'''select
           e.name,
           count(v.id)
           from public.employers e
               join public.vacancies v on e.id=v.employer_id
           group by 1
           order by 2 desc
           limit {n}
         '''

# Получаем результаты запроса сразу в виде датафрейма
df_lider_employer = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_lider_employer)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\1186734325.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_lider_employer = pd.read_sql_query(query, connection)


,name,count
0,Яндекс,1933
1,Ростелеком,491
2,Тинькофф,444
3,СБЕР,428
4,Газпром нефть,331


Вывод по заданию 5.1. Первое место по количеству опубликованных вакансий занимает компания Яндекс, пятое место по количеству опубликованных вакансий занимает компания Газпром нефть. При этом Яндекс занимает 6 место в рейтинге топ компаний по версии HeadHunter и относящиеся к крупнейшим компаниям.

Задания 5.2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём. 
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей. Впишите его название в поле ниже в том виде, который вернул запрос.

In [13]:
n = 1
query = f'''select a.name,
           count(e.id) employer,
           count(v.id) vacancies
    from areas a
    left join vacancies v on a.id = v.area_id
    left join employers e on a.id = e.area
    group by a.name
    HAVING count(v.id) = 0 
    order by 2 DESC
    limit {n}
        '''
# Получаем результаты запроса сразу в виде датафрейма
df_employer_luser_area = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_employer_luser_area)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\132281874.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employer_luser_area = pd.read_sql_query(query, connection)


,name,employer,vacancies
0,Россия,410,0


Вывод по заданию 5.2. Россия является регионом с самым большим количеством работодателей при отсутсвии опубликованных вакансий, что указывает косвенно на то, что Россия указана как регион даже если расположение работодателя (вакансии) находится в субъекте РФ, так как дополнительный вывод результатов (топ 10) показал, что в топ-10 входят также регионы РФ. 

Задание 5.3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. 
Выберите максимальное значение из получившегося списка.

In [14]:
#Задаем ограничение вывод результатов
n = 1
query = f'''select  
           e.name,
           count(distinct v.area_id) area
           from public.employers e
                join public.vacancies v on e.id=v.employer_id
           group by e.name
           order by 2 desc
           limit {n}
         '''
# Получаем результаты запроса сразу в виде датафрейма
df_employer_count_area = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_employer_count_area)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\1837352232.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employer_count_area = pd.read_sql_query(query, connection)


,name,area
0,Яндекс,181


Вывод по заданию 5.3. Яндекс является компанией, которая представлена в самом большом количестве регионов.

Задание 5.4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. Введите количество, которое вернул запрос, в поле ниже.

In [15]:
query = f'''select  
            count(e.name)
                        from public.employers e
                left join public.employers_industries e_i on e_i.employer_id=e.id
                left join public.industries i on i.id=e_i.industry_id
            where i.id is null or i.name is null
         '''

# Получаем результаты запроса сразу в виде датафрейма
df_employer_industries = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_employer_industries)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\2256145711.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employer_industries = pd.read_sql_query(query, connection)


,count
0,8419


Вывод по заданию 5.4. Для 8419 компаний не указана сфера деятельности, что составляет почти треть всех работодателей, занесенных в базу. Можно предположить, что либо это крупные организации чье направление деятельности охватывает различные сектора, либо это маленькие частные организации, которые не в полной мере ответсвенно отнеслись к оформлению и публикации вакансий.

Задание 5.5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. Введите в поле ниже название этой компании так же, как оно указано в результате запроса.

In [16]:
#Задаем ограничение вывода, топ-3
n = 3
query = f'''select  
            e.name name_company,
            count(i.name) name_industry,
            count(i.id) id_industry
                        from public.employers e
                left join public.employers_industries e_i on e_i.employer_id=e.id
                left join public.industries i on i.id=e_i.industry_id
            group by e.name
            having   count(i.name) = 4 or count(i.id) = 4
            order by 1 asc
            limit {n}
        '''

# Получаем результаты запроса сразу в виде датафрейма
df_employer_count_industries = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_employer_count_industries)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\1062859363.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employer_count_industries = pd.read_sql_query(query, connection)


,name_company,name_industry,id_industry
0,101 Интернет,4,4
1,21vek.by,4,4
2,2ГИС,4,4


Вывод по задаче 5.5. Компания 2ГИС занимает 3 место среди организаицй для которых указаны 4 сферы деятельности и отсортированные в алфавитном порядке.

Задание 5.6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».

In [ ]:
query = f'''select  
            count(e.name) 
                        from public.employers e
                left join public.employers_industries e_i on e_i.employer_id=e.id
                left join public.industries i on i.id=e_i.industry_id
            where i.name = 'Разработка программного обеспечения'
            
        '''
# Получаем результаты запроса сразу в виде датафрейма
df_employer_count_PO = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_employer_count_PO)

C:\Users\user\AppData\Local\Temp\ipykernel_13680\551664993.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employer_count_PO = pd.read_sql_query(query, connection)


,count
0,3553


Вывод по задаче 5.6. Для 3553 компаний в качестве сферы деятельности указана "Разработка программного обеспечения", что составляет седьмую часть от всех организаций, которые опубликовали свои вакансии.

Задание 5.7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании, собранных в этой таблице.

In [19]:
query = f'''(select  
            a.name,
            count(v.id) count_vacansies
            from public.vacancies v
                left join employers e on e.id = v.employer_id
                left join public.areas a on v.area_id=a.id
            where e.name = 'Яндекс'   
            group by a.name
            having (a.name = 'Москва' or a.name = 'Санкт-Петербург' or a.name = 'Новосибирск' or a.name = 'Екатеринбург' or a.name = 'Казань' or a.name = 'Красноярск' or a.name = 'Нижний Новгород' or a.name = 'Челябинск'
            or a.name = 'Уфа' or a.name = 'Краснодар' or a.name = 'Самара' or a.name = 'Ростов-на-Дону' or a.name = 'Омск' or a.name = 'Воронеж' or a.name = 'Пермь' or a.name = 'Волгоград')
            order by 2 desc)
            
            union all

            (select
                 'total',
                 count(v.id)
                 from public.vacancies v
                      left join employers e on v.employer_id = e.id
                      left join public.areas a on v.area_id=a.id
                where e.name = 'Яндекс' and (a.name = 'Москва' or a.name = 'Санкт-Петербург' or a.name = 'Новосибирск' or a.name = 'Екатеринбург' or a.name = 'Казань' or a.name = 'Красноярск' or a.name = 'Нижний Новгород' or a.name = 'Челябинск'
            or a.name = 'Уфа' or a.name = 'Краснодар' or a.name = 'Самара' or a.name = 'Ростов-на-Дону' or a.name = 'Омск' or a.name = 'Воронеж' or a.name = 'Пермь' or a.name = 'Волгоград')
            )
             
        '''
# Получаем результаты запроса сразу в виде датафрейма
df_yandex_area_count = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_yandex_area_count)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\3963243763.py:27: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_yandex_area_count = pd.read_sql_query(query, connection)


,name,count_vacansies
0,total,485
1,Москва,54
2,Санкт-Петербург,42
3,Екатеринбург,39
4,Нижний Новгород,36
5,Новосибирск,35
6,Воронеж,32
7,Краснодар,30
8,Самара,26
9,Уфа,26


Вывод по заданию 5.7. Наибольшее количество вакансий от компании Яндекс опубликованы в Москве. При этом количество опубликованных вакансий не зависит от количетсва проживающих людей в городе-миллионике, так как порядок вывода городов миллионников по убыванию количества опубликованных вакансий не соответсвует перечню данных городов в порядке убывания по количеству проживающих людей. Суммарно в городах-миллионниках опубликовано 485 вакансий, что составляет 4 часть от всех вакансий компании Яндекс.

Задание 6.1. Сколько вакансий имеет отношение к данным?

In [18]:
query = f'''select 
             count(distinct v.id)
             from public.vacancies v
             where lower(v.name) like '%data%' or lower(v.name) like '%данн%' 
           '''

# Получаем результаты запроса сразу в виде датафрейма
df_vacan_data = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_vacan_data)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\3759740236.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacan_data = pd.read_sql_query(query, connection)


,count
0,1771


Вывод по заданию 6.1. 1771 опубликованная вакансия имеет отношение к данным, то есть 3,5% от всех опубликованных вакансий.

Задание 6.2. Сколько есть подходящих вакансий для начинающего дата-сайентиста?

In [ ]:
query = f'''select 
              count(v.name)
from vacancies v

WHERE
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)

    AND (v.name ILIKE '%junior%'
    OR v.experience ILIKE '%Нет опыта%'
    OR v.employment ILIKE '%Стажировка%')

           '''

# Получаем результаты запроса сразу в виде датафрейма
df_vacan_data_jun = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_vacan_data_jun)

C:\Users\user\AppData\Local\Temp\ipykernel_13680\840695076.py:27: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacan_data_jun = pd.read_sql_query(query, connection)


,count
0,51


Выводы по заданию 6.2. Для начинающих спечиалистов по дата-сайентс опубликована всего 51 вакансия, что составляет 2,8% от всех опубликованных вакансий, имеющих отношения к данным. Не так-то просто найти начинающему специалисту работу. А вот опытных специалистов ищуют активно.

Задание 6.3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

In [20]:
query = f'''select 
              count(v.name)
from vacancies v

WHERE
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%'))
    AND (key_skills ILIKE '%SQL%' or key_skills ILIKE '%postgres%')
    '''

# Получаем результаты запроса сразу в виде датафрейма
df_vacan_data_DS = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_vacan_data_DS)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\272405666.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacan_data_DS = pd.read_sql_query(query, connection)


,count
0,229


Вывод 6.3. В 229 вакансиях в качестве ключевого навыка указан SQL или postgres, чт о составляет почти 8 часть, от всех опубликованных вакансий, имеющих отношение к данным.

Задание 6.4. С помощью запроса, аналогичного предыдущему, проверьте, насколько популярен Python в требованиях работодателей к DS. Вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

In [ ]:
query = f'''select 
              count(v.name)

from vacancies v

WHERE
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%'))
    AND (key_skills ILIKE '%Python%')
    '''

# Получаем результаты запроса сразу в виде датафрейма
df_vacan_data_Python = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_vacan_data_Python)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\3883769894.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacan_data_Python = pd.read_sql_query(query, connection)


,count
0,357


Вывод по заданию 6.4. В 357 вакансиях в качестве ключевого навыка указан Python, чт о составляет почти 5 часть, от всех опубликованных вакансий, имеющих отношение к данным.

Задание 6.5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?

In [ ]:

query = f'''SELECT 
ROUND(AVG(CHAR_LENGTH(key_skills) - CHAR_LENGTH(replace(key_skills, CHR(9), '')) + 1), 2) AS "Слов"

FROM vacancies
WHERE
    (lower(name) LIKE '%data scientist%' OR
    lower(name) LIKE '%data science%' OR
    lower(name) LIKE '%исследователь данных%' OR
    (name LIKE '%ML%' AND lower(name) NOT LIKE'%html%') OR
    lower(name) LIKE '%machine learning%' OR
    lower(name) LIKE '%машинн%обучен%') AND
    (key_skills IS NOT NULL AND key_skills != '')
'''

# Получаем результаты запроса сразу в виде датафрейма
df_vacan_data_keys = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_vacan_data_keys)

C:\Users\user\AppData\Local\Temp\ipykernel_13680\2913089912.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacan_data_keys = pd.read_sql_query(query, connection)


,Слов
0,6.41


Вывод по заданию 6.5. В среднем указаным 6,41 ключевой навык для вакансий, имеющих отношение к данным.

Задание 6.6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля experience).

In [22]:
query = f'''SELECT 
            v.experience, ROUND(AVG(coalesce((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0)

            from vacancies v

            WHERE
            (
                   (v.name ILIKE '%data scientist%')
                OR (lower(v.name) ILIKE '%data science%')
                OR (v.name ILIKE '%исследователь данных%')
                OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
                OR (v.name ILIKE '%machine learning%')
                OR (v.name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL)
            )

            GROUP BY v.experience
            ORDER BY 1'''


# Получаем результаты запроса сразу в виде датафрейма
df_vacan_data_exp = pd.read_sql_query(query, connection)

# Выводим результаты запроса
display(df_vacan_data_exp)

C:\Users\user\AppData\Local\Temp\ipykernel_21916\2455904123.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacan_data_exp = pd.read_sql_query(query, connection)


,experience,round
0,Более 6 лет,NaN
1,Нет опыта,74643.0
2,От 1 года до 3 лет,145236.0
3,От 3 до 6 лет,243115.0


Вывод по заданию 6.6. В среднем, зарплата для дата-сайентиста с опытом работы от 3 до 6 лет составляет 243115 руб, что в 3,25 раза больше, чем для специалистов без опыта работы. При этом для специалистов с опытом работы более 6 лет зарплатная вилка не указана, что косвенно указывает на факт объема знаний опытных специалистов и индивидуального подхода к предложениям зарплаты для них.

Общий вывод по результатам анализа базы вакансий. 
В среднем на каждого работодателя (23501) в базе приходится по 2.09 вакансий (49197). База вакансий HeadHunter охватывает вакансии по 1362 регионам и по 294 сферам деятельности. Только у половины вакансий указана заработная плата, которая в среднем по нижней границе составляет 71065 руб, и в среднем по верхней границе составляет 110537 руб. Самое большое количество вакансий с полной занятостью и полным рабочим днем, на стором месте с полной занятостью и удаленной работой. Самое количество вакансий опубликовано с требуемцым опытом работы от 1 года до 3х лет. При этом самое низкое количество вакансий опубликовано для поиска специалистов с опытом работы более 6 лет. Самый большой охват вакансий в разрезе регионов у компании Яндекс (181). Разработка программного обеспечения в качестве сферы деятельности указаны у 3553 вакансий, т.е. у 7 части от общего количества всех работодателей. 1771 опубликованная вакансия имеет отношение к данным, то есть 3,5% от всех опубликованных вакансий, а средняя зарплата для данных специалистов с опытом работы от 3 до 6 лет в 2,3 раза больше средней зарплаты по верхней границы для всей базы вакансий, а для специалистов без опыта работы средняя заработная плата указана больше (74643), чем средняя заработная плата по нижней границе для всей базы (71065).